packages/graphql-client/src/graphql-client/graphql-client.ts

```diff
@@ -165,3 +184,276 @@ function generateRequest(
     }
   };
 }
+
+async function* getStreamBodyIterator(
+  response: Response,
+): AsyncIterableIterator<string> {
+  // Support node-fetch format
+  if ((response.body as any)![Symbol.asyncIterator]) {
+    for await (const chunk of response.body! as any) {
+      yield (chunk as Buffer).toString();
+    }
+  } else {
+    const reader = response.body!.getReader();
+    const decoder = new TextDecoder();
+
+    let readResult: ReadableStreamReadResult<DataChunk>;
+    try {
+      while (!(readResult = await reader.read()).done) {
+        yield decoder.decode(readResult.value);
+      }
+    } finally {
+      reader.cancel();
+    }
+  }
+}
+
+function readStreamChunk(
+  streamBodyIterator: AsyncIterableIterator<string>,
+  boundary: string,
+) {
+  return {
+    async *[Symbol.asyncIterator]() {
+      try {
+        let buffer = "";
+
+        for await (const textChunk of streamBodyIterator) {
+          buffer += textChunk;
+
+          if (buffer.indexOf(boundary) > -1) {
+            const lastBoundaryIndex = buffer.lastIndexOf(boundary);
+            const fullResponses = buffer.slice(0, lastBoundaryIndex);
+
+            const chunkBodies = fullResponses
+              .split(boundary)
+              .filter((chunk) => chunk.trim().length > 0)
+              .map((chunk) => {
+                const body = chunk
+                  .slice(
+                    chunk.indexOf(HEADER_SEPARATOR) + HEADER_SEPARATOR.length,
+                  )
+                  .trim();
+                return body;
+              });
+
+            if (chunkBodies.length > 0) {
+              yield chunkBodies;
+            }
+
+            buffer = buffer.slice(lastBoundaryIndex + boundary.length);
+
+            if (buffer.trim() === `--`) {
+              buffer = "";
+            }
+          }
+        }
+      } catch (error) {
+        throw new Error(
+          `Error occured while processing stream payload - ${getErrorMessage(
+            error,
+          )}`,
+        );
+      }
+    },
+  };
+}
+
+function generateRequestStream(
+  fetch: ReturnType<typeof generateFetch>,
+): GraphQLClient["requestStream"] {
+  return async (...props) => {
+    if (!DEFER_OPERATION_REGEX.test(props[0])) {
+      throw new Error(
+        formatErrorMessage(
+          "This operation does not result in a streamable response - use request() instead.",
+        ),
+      );
+    }
+
+    try {
+      const response = await fetch(...props);
+
+      const { status, statusText } = response;
+
+      if (!response.ok) {
+        throw new Error(statusText, { cause: response });
+      }
+
+      const responseContentType = response.headers.get("content-type") || "";
+      const isNotSupportedContentType = Object.values(CONTENT_TYPES).every(
+        (type) => !responseContentType.includes(type),
+      );
+
+      if (isNotSupportedContentType) {
```

Comment:
Personal preference, but I wonder if it would make this easier to read if we went with something like this:

```ts
switch (true) {
  responseContentType.includes(CONTENT_TYPES.json):
    return createJsonAsyncIterator();
  responseContentType.some((type) => CONTENT_TYPES.includes(type)):
    return createMultipartAsyncInterator();
  default:
    throw
}
```

to create a little bit more separation between them. I feel like there might be a couple of different method concerns joined into this one.